shim/src/scopeHandler.js

```diff
@@ -1,128 +1,142 @@
-  // climbing the scope chain and eventually reaching the unsafeGlobal
-  // object, which is bad.
-
-  // note: unscopables! every string in Object[Symbol.unscopables]
-
-  // todo: we'd like to just have has() return true for everything, and then
-  // use get() to raise a ReferenceError for anything not on the safe global.
-  // But we want to be compatible with ReferenceError in the normal case and
-  // the lack of ReferenceError in the 'typeof' case. Must either reliably
-  // distinguish these two cases (the trap behavior might be different), or
-  // we rely on a mandatory source-to-source transform to change 'typeof abc'
-  // to XXX. We already need a mandatory parse to prevent the 'import' and
-  // 'import.meta' expressions, since they're special forms instead of merely
-  // being a global variable
-
-  // note: if we make has() return true always, then we must implement a
-  // set() trap to avoid subverting the protection of strict mode (it would
-  // accept assignments to undefined globals, when it ought to throw
-  // ReferenceError for such assignments)
-
-  has(target, prop) {
-    // proxies stringify 'prop', so no TOCTTOU danger here
-    if (prop === 'eval') {
+    },
+
+    // eslint-disable-next-line class-methods-use-this
+    set(target, prop, value) {
+      // Set the value on the shadow. The target itself is an empty
+      // object that is only used to prevent a frozen eval property.
+      // todo: use this.shadowTarget, for speedup
+
+      // new todo: allow modifications when target.hasOwnProperty(prop) and it
+      // is writable, assuming we've already rejected overlap (see
+      // createSafeEvaluatorFactory.factory). This TypeError gets replaced with
+      // target[prop] = value
+      if (objectHasOwnProperty(target, prop)) {
+        // todo: shim integrity: TypeError, String
+        throw new TypeError(`do not modify endowments like ${String(prop)}`);
+      }
+      getPrototypeOf(target)[prop] = value;
+      // Return true after successful set.
       return true;
-    }
-    if (prop === 'arguments') {
+    },
+
+    // we need has() to return false for some names to prevent the lookup  from
+    // climbing the scope chain and eventually reaching the unsafeGlobal
+    // object, which is bad.
+
+    // note: unscopables! every string in Object[Symbol.unscopables]
+
+    // todo: we'd like to just have has() return true for everything, and then
+    // use get() to raise a ReferenceError for anything not on the safe global.
+    // But we want to be compatible with ReferenceError in the normal case and
+    // the lack of ReferenceError in the 'typeof' case. Must either reliably
+    // distinguish these two cases (the trap behavior might be different), or
+    // we rely on a mandatory source-to-source transform to change 'typeof abc'
+    // to XXX. We already need a mandatory parse to prevent the 'import' and
+    // 'import.meta' expressions, since they're special forms instead of merely
```

Comment:
Stale comment. I was wrong about `import.meta`. We need the parse only to prevent the `import` expression.